### PR TITLE
LiteralOrSource::Stack

### PR DESCRIPTION
**Pull request by ModProg**

### CODE REVIEW

File: budvm/src/lib.rs [hunk 1]
```diff
@@ -2111,14 +2119,33 @@ where
...
 
-    fn resolve_value_or_source<'v>(
+    fn resolve_value_or_source_popped<'v>(
+        &'v mut self,
+        value: &'v ValueOrSource,
+    ) -> Result<Cow<'v, Value>, FaultKind> {
+        match value {
+            ValueOrSource::Argument(index) => self.resolve_argument(*index).map(Cow::Borrowed),
+            ValueOrSource::Variable(index) => self.resolve_variable(*index).map(Cow::Borrowed),
+            ValueOrSource::Value(value) => Ok(value).map(Cow::Borrowed),
```
Comment (ModProg) [inline budvm/src/lib.rs:2125]: We could use `resolve_value_or_source_non_mut` but I don't feel like it improves readability.

File: budvm/src/lib.rs [hunk 2]
```diff
@@ -2651,6 +2685,8 @@ macro_rules! checked_op {
                         ))
                     }
                 };
+                // TODO handle failures
+                self.stack.pop_n(to_pop);
```
Comment (ModProg) [inline budvm/src/lib.rs:2686]: This function can exit on errors early, do we need to clean up the stack in these cases?

File: budvm/src/lib.rs [hunk 3]
```diff
@@ -2186,6 +2214,8 @@ where
             })?,
         };
 
+        self.stack.pop_n(to_pop);
```
Comment (ModProg) [inline budvm/src/lib.rs:2213]: This function can exit on errors early, do we need to clean up the stack in these cases?

File: budvm/src/lib.rs [hunk 4]
```diff
@@ -3574,7 +3609,18 @@ impl Stack {
     #[inline]
     pub fn top(&self) -> Result<&Value, FaultKind> {
         if self.length > 0 {
-            Ok(&self.values[self.length])
+            Ok(&self.values[self.length - 1])
```
Comment (ModProg) [inline budvm/src/lib.rs:3610]: not actually using top anymore, but this caused out of bounds errors before :D

**Review by ecton — APPROVED** [review]
This looks great, thank you. As per Discord, one question was about whether we should care about popping the values from the stack in the event of a fault. Here's my response on Discord:

>  I decided that when a fault occurs, its like a program on the computer died. You may want to inspect the stack to understand why a fault occurred, so leaving the stack intact with when the fault happens allows you to do that analysis. In fact, this pop-after-the-op-is-done approach helps with that since even those transient values will still be present on the stack if the operation itself fails.

Thank you again for tackling this, I like how it turned out!

**Pull request merged**